watcher.go

```diff
@@ -0,0 +1,141 @@
+package consul
+
+import (
+	"context"
+	"errors"
+	"net"
+	"net/http"
+	"strconv"
+	"time"
+)
+
+// WatcherFunc is the function signature for the callback from watch.  It
+// passes a list of KeyData representing the most recent value or values stored
+// at the key or everything below the prefix. It can be nil.  error is passed
+// through to the handler if it is not temporary. Caller is responsible for
+// stopping the watch if error is determined to be unrecoverable
+type WatcherFunc func([]KeyData, error)
+
+// Watcher is the struct upon which Watch and WatchPrefix are built.
+type Watcher struct {
+	Client *Client
+	// MaxAttempts limits the number of subsequent failed API calls before
+	// bailing out of the watch.
+	MaxAttempts int
+}
+
+var (
+	DefaultWatcher = &Watcher{}
+
+	// ErrMaxAttempts gets passed to the handler when the number of subsequent
+	// attempts reaches Watcher.MaxAttempts
+	ErrMaxAttempts = errors.New("watcher: max attempts reached. giving up")
+)
+
+func (w *Watcher) client() *Client {
+	if client := w.Client; client != nil {
+		return client
+	}
+	// This is copied from DefaultTransport.
+	// we don't do an actual copy because the Trasnport uses mutexes.
+	// copied mutexes are ineffective.
+	// You will panic on concurrent map access if running
+	// with other clients without instantiating a new transport here.
+	ts := &http.Transport{
```

Comment:
Any reason not to make this a global variable? `http.Transport` manages a connection pool, it's designed to be shared.

Reply:
We could.  `WatcherTransport`?

Reply:
👍